libs/s25main/mapGenerator/ValueMap.h

```diff
@@ -0,0 +1,174 @@
+// Copyright (c) 2005 - 2017 Settlers Freaks (sf-team at siedler25.org)
+//
+// This file is part of Return To The Roots.
+//
+// Return To The Roots is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 2 of the License, or
+// (at your option) any later version.
+//
+// Return To The Roots is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Return To The Roots. If not, see <http://www.gnu.org/licenses/>.
+
+#pragma once
+
+#include "world/NodeMapBase.h"
+#include <cmath>
+
+namespace rttr { namespace mapGenerator {
+
+    template<typename T_Value>
+    struct ValueRange
+    {
+        ValueRange(T_Value minimum, T_Value maximum) : minimum(minimum), maximum(maximum) {}
+
+        const T_Value minimum;
+        const T_Value maximum;
+
+        /**
+         * Computes the difference between maximum and minimum value.
+         *
+         * @returns the difference between maximum and minimum.
+         */
+        T_Value GetDifference() const { return maximum - minimum; }
+    };
+
+    /**
+     * Maps the specified value of a range between the minimum and maximum to the corresponding index of an array
+     * of the specified size. It's important to note, that this function make sure that following ratios are equal:
+     * v / (max - min) = MapRangeToIndex(v, min, max, size) / size
+     *
+     * @param value value to map to the new range
+     * @param range minimum and maximum value
+     * @param size size of the container to find an index for
+     *
+     * @returns an index for a container which corresponds to the specified value within the specified range.
+     */
+    template<typename T_Value>
+    unsigned MapValueToIndex(T_Value value, const ValueRange<T_Value>& range, size_t size)
+    {
+        T_Value difference = range.GetDifference();
+
+        if(difference)
+        {
+            double slope = 1. * (size - 1) / difference;
+            return static_cast<unsigned>(round(slope * (value - range.minimum)));
+        }
+
+        return 0u;
+    }
+
+    template<typename T_Value>
+    class ValueMap : public NodeMapBase<T_Value>
```

Comment:
This looks like it could be free functions and NodeMapBase be final again, or what was the reason?

Reply:
Yes agreed & fixed!